other/no-loss-lottery/lottery/src/lib.rs

```diff
@@ -0,0 +1,321 @@
+            let staking: Staking = staking_address.into();
+            // staking user badge
+            let staking_badge = staking.new_user();
+
+            let component = Self {
+                staking,
+                admin_badge: admin_resource_def, 
+                ticket_minter: Vault::with_bucket(ticket_minter),
+                staking_user: Vault::with_bucket(staking_badge),
+                staking_address,
+                withdraw_vault: Vault::new(RADIX_TOKEN),
+                players: HashMap::new(),
+                ticket_id_counter: 0,
+                tickets: tickets,
+                minted: Vec::new(),
+                ticket_price: price,
+                lottery_epoch: 0,
+                lottery_name: String::new(),
+                ended: HashMap::new(),
+            }
+            .instantiate();
+            
+            (component, admin_badge)
+        }
+
+        #[auth(admin_badge)]
+        pub fn start_lottery(&mut self, name: String, epoch: u64) {
+            assert!(
+                self.lottery_epoch == 0 || Context::current_epoch() > self.lottery_epoch, 
+                "Cannot initiate new lottery when current one is not ended yet"
+            );
+
+            assert!(epoch > 0, "Epoch cannot be zero");
+            assert!(!name.is_empty(), "Name cannot be empty");
+
+            self.lottery_epoch = epoch;
+            self.lottery_name = name
+        }
+
+        #[auth(admin_badge)]
+        pub fn end_lottery(&mut self) {
+            assert!(Context::current_epoch() > self.lottery_epoch, "Lottery is not ended yet");
+
+            debug!("++++++Lottery Finished++++++");
+
+            // withdraw staking
+            self.staking_user.authorize(|auth| {
+                let withdraw = self.staking.withdraw(auth);
+                self.withdraw_vault.put(withdraw);
+            });            
+
+            // pick winner
+            // take hash
+            let hash = Context::transaction_hash().to_string();
+            //take first 5 bits of result hash
+            let seed = &hash[0..5];
+            //convert 10 hex bits to usize
+            let result = usize::from_str_radix(&seed, 16).unwrap();
+            // get winner index
+            let index = result % self.minted.iter().count();
+            // get winner ticket id
+            let winner = self.minted[index];
+            debug!("Winner ID: {}", winner);
+            // save result
+
+            self.staking_user.authorize(|auth| {
+                let reward = self.staking.get_reward(auth);
+                debug!("Reward: {}", reward.amount());
+                self.ended.insert(self.lottery_name.to_string(), (winner, Vault::with_bucket(reward)));
+            });
+                        
+            // clear minted array
+            self.minted.clear();
+            debug!("=======================");
+        }
+
+        pub fn buy_ticket(&mut self, player: Address, payment: Bucket) -> (Bucket, Bucket) {
```

Comment:
Storing the stake data inside the NFT would allow you to simplify your code. You would not need to require a player_address in the params of this method